### ISSUE TITLE
Overlap/merge with xarrayutils.vertical_coordinate

### ISSUE THREAD

**Issue by jbusecke** [raphaeldussin/VCR#1]
Hey, 

just stumbled across this package by chance. I just wanted to point out that I am developing something very similar over [here](https://github.com/jbusecke/xarrayutils/blob/master/xarrayutils/vertical_coordinates.py) (usage examples can be found [here](https://github.com/jbusecke/xarrayutils/blob/master/doc/vertical_coords.ipynb)). My implementation is more xarray focused and so maybe the overlap is not that big of a deal? But we are using the same method at the core I think.

Just wanted to drop this here so if we plan to develop these further we can possibly collaborate.

Cheers.

**Comment by raphaeldussin** [comment]
hey julius! wish I saw your repo beforehand. I couldn’t find a conservative 1d anywhere so I wrote my own. That’s the kind of functionality I’d like to have in xarray or xgcm. Do you want to work on a PR together? 

cheers,
raf

**Comment by jbusecke** [comment]
Should we get together some time this or next week and chat about it? Would love to collaborate!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by raphaeldussin** [comment]
what about before/after the next regional MOM6 check-in ?

**Comment by jbusecke** [comment]
I am not necessarily going I think. But when is it?

Also Ill cc @cspencerjones here. He has been developing [xlayers](https://github.com/cspencerjones/xlayers). Ultimately I would love it if we can somehow have a library of all different approaches and combine them to the ultimate vertical regrid/remap package. Either way it would be great to test and benchmark these approaches against each other!

**Comment by raphaeldussin** [comment]
I figure there's going to be one in the next 2 weeks but we can do another time no worries.

I'm wondering what's the most useful approach: xarray has already an interp 1d function with several methods. Adding a 1d conservative method might be useful for other purposes maybe?
e.g. some flow rate interpolated in time?

Or if not, I think xgcm would make the most sense. btw I like the new metrics functions ;)

**Comment by jbusecke** [comment]
We have talked about that issue for quite a bit over at xgcm (see [here](https://github.com/xgcm/xgcm/issues/144) and [here](https://github.com/xgcm/xgcm/issues/68)). I am not sure where @rabernat stands on this these days, but I am torn...on the one hand, one doesn't really need the c-grid information for these operations, and this goes far beyond the simple diff/interp functionality of xgcm. But on the other, the cell boundary information is neatly encoded in the xgcm data model. It would be adding a substantial amount of code to xgcm, so perhaps a stand-alone package would be better?

> btw I like the new metrics functions ;)

Aww nice. Stoked that you are using it!

**Comment by jbusecke** [comment]
> I'm wondering what's the most useful approach: xarray has already an interp 1d function with several methods. 

The 1d interpolation of xarray does not currently support 'reverse' interpolation. E.g. it easy to find the temperature as a function of dimension (temp@30m), but the inverse is not possible (depth@20deg) with pure xarray. This is a key part of any regridding/remapping based on tracer coordinates. Not sure if this case is general enough for xarray?

> Adding a 1d conservative method might be useful for other purposes maybe?
e.g. some flow rate interpolated in time?

I am not sure I understand this problem. How would you use the conservative remapping here?

**Comment by cspencerjones** [comment]
I would definitely be up for some benchmarking if we can agree on what we should actually be comparing. If you guys talk then I could zoom in (though not on monday or tuesday of next week)?

**Comment by jbusecke** [comment]
I have some tests in [xarrayutils](https://github.com/jbusecke/xarrayutils/blob/master/xarrayutils/test/test_vertical_coordinates.py). These might be a good starting point. 

I would envision a package that works to implement regridding/remapping in a bunch of different ways. We could then choose different 'engines', like xlayers, numba, or pure xarray (default for simplicity), but obviously these should all be resulting in the same thing hehe.

**Comment by cspencerjones** [comment]
I think we'll have to negotiate about that once we know the stats. I am not really up for maintaining another large package and xlayers does a very specific thing well and might work well with standalone documentation. Also it has different dependencies than your package (that's kind of the point, right?).

**Comment by jbusecke** [comment]
Indeed, the testing comes first. I was also not necessarily suggesting that we dissolve and merge all components. As you say xlayers is quite self-contained, but another more general package could have it as an optional dependency (which would be nice if it ends up beating all other implementations on speed 😁 ). 

Id be happy to try to construct a benchmark with a large synthetic or cloud dataset once you think docs and code are ready on the xlayer side. 

> Also it has different dependencies than your package (that's kind of the point, right?).

Yes exactly. This was the reason why I constructed a pure python implementation.

**Comment by raphaeldussin** [comment]
I get my 3d remapping done under a second. Constructing the regridding matrix is probably where the time is mostly spent. Since the remapping matrix is going to be sparse, I test first for non-overlapping cells which sets most of them to zero without additional numerical operations. Computing the overlap is only 4 operations...

When applying the weights, I use np.matmul and it's blazing fast. There's some transpositions needed but it doesn't seem to cost much.

I'd have it implemented for xarray with apply_ufunc or map_blocks and call it a day :)

you can test out this notebook (all data pulled from opendap):

https://github.com/raphaeldussin/VCR/blob/master/examples/Production_remapping_WOA2013.ipynb

**Comment by jbusecke** [comment]
Ok that seems much more efficient than my very simple implementation. Ill take a closer look soon. Very keen on merging our efforts here! Are you on campus some time this week?

**Comment by raphaeldussin** [comment]
I think your implementation is very similar algorithmically (there are not a 100 ways of doing it :) )
but you compute overlap for all vertical cells and then zero some of them out. Not sure about the performance of xr.dot product. I trust Ryan's method of writing a fast numpy kernel then add the xarray/dask layer on top. I don't go much on campus, but I'm sure you can lure me to small world sometime next week ;)

**Comment by jbusecke** [comment]
> I think your implementation is very similar algorithmically (there are not a 100 ways of doing it :) )

Well, there is always a full on for-loop hahahaha.

> you compute overlap for all vertical cells and then zero some of them out

It seemed very fast to me but I think your logic is a great addition to save computation.

> I trust Ryan's method of writing a fast numpy kernel then add the xarray/dask layer on top.

That is a great way to do it. But the way it is implemented now, does is work for chunks with more than one dimension? If not this will get slowed down significantly by the need to wrap in np.vectorize (see my interpolation methods). In that case, I believe you would lose the speed advantage, particularly when working with time-space varying depth coordinates. 

> but I'm sure you can lure me to small world sometime next week ;)

😆 If you like join us in the ivy realm. We can have a nice lunch too. @gmacgilchrist might be interested in joining the fun too?

**Comment by gmacgilchrist** [comment]
Certainly keen to follow the outcome of all this. I'd be happy to join. I can set up some good watermass transformation test cases to ensure that the approach is sufficiently accurate to perform large-scale calculations requiring closed budgets.
